src/Stratis.Bitcoin/BlockPulling2/BlockPullerBehavior.cs

```diff
@@ -0,0 +1,217 @@
+        public void OnIbdStateChanged(bool isIbd)
+        {
+            this.logger.LogTrace("()");
+
+            // Recalculates the max samples count that can be used for quality score calculation.
+            int samplesCount = isIbd ? IbdSamplesCount : NormalSamplesCount;
+            this.averageSizeBytes.SetMaxSamples(samplesCount);
+            this.averageDelaySeconds.SetMaxSamples(samplesCount);
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Recalculates the quality score for this peer.</summary>
+        /// <param name="bestSpeedBytesPerSecond">Speed in bytes per second that is considered to be the maximum speed.</param>
+        public void RecalculateQualityScore(int bestSpeedBytesPerSecond)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(bestSpeedBytesPerSecond), bestSpeedBytesPerSecond);
+
+            this.QualityScore = (double)this.SpeedBytesPerSecond / bestSpeedBytesPerSecond;
+
+            if (this.QualityScore < MinQualityScore)
+                this.QualityScore = MinQualityScore;
+
+            if (this.QualityScore > MaxQualityScore)
+                this.QualityScore = MaxQualityScore;
+
+            this.logger.LogTrace("Quality score was set to {0}.", this.QualityScore);
+            this.logger.LogTrace("(-)");
+        }
+        
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(message), message.Message.Command);
+
+            if (message.Message.Payload is BlockPayload block)
+            {
+                uint256 blockHash = block.Obj.GetHash();
+
+                this.logger.LogTrace("Block '{0}' delivered.", blockHash);
+                this.blockPuller.PushBlock(blockHash, block.Obj, peer.Connection.Id);
+            }
+
+            this.logger.LogTrace("(-)");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>Requests blocks from this peer.</summary>
+        /// <param name="hashes">Hashes of blocks that should be asked to be delivered.</param>
+        /// <exception cref="OperationCanceledException">Thrown in case peer is in the wrong state or TCP connection was closed during sending a message.</exception>
+        public async Task RequestBlocksAsync(List<uint256> hashes)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(hashes), hashes.Count);
+
+            var getDataPayload = new GetDataPayload();
+
+            foreach (uint256 uint256 in hashes)
+            {
+                var vector = new InventoryVector(InventoryType.MSG_BLOCK, uint256);
+                vector.Type = this.AttachedPeer.AddSupportedOptions(vector.Type);
+
+                getDataPayload.Inventory.Add(vector);
+            }
+
+            if ((this.AttachedPeer == null) || (this.AttachedPeer.State != NetworkPeerState.HandShaked))
```

Comment:
Attached peer must not be null, so you don't need to check for that

Reply:
Fixed